Add median and relatives to std/stats

HJarausch opened the pull request:
The statistics module is missing the computation of the median of a sequence.
This PR adds
**quickSelect** : quick selecting the  k-th smallest or k-th largest element of a sequence using the _Lomuto_ partition scheme

For a sequence of odd length ``N``  the **median** is that element which is larger-or-equal to ``(N-1)/2)`` elements and greater-or-equal than ``(N-1)/2)`` elements of the sequence.

**median_low** computes the **median** for a sequence of odd length. In case of even length ``N``, ``median_low`` has the property that ``N/2`` elements are less than or equal this value   and ``N/2+1`` elements are greater than or equal this value.
It can be computed for any sequence of _comparable items_.

**median_high** computes the **median** for a sequence of odd length. In case of even length ``N``, ``median_high`` has the property that ``N/2`` elements are greater than or equal this value   and ``N/2+1`` elements are less than or equal this value.
It can be computed for any sequence of _comparable items_.

**median** computes the ``median`` of a sequence of _numeric items_. For a sequence of odd length it is equivalent to ``median_low`` and ``median_high``.  In case of a sequence of even length it computes the arithmetic mean of  ``median_low`` and ``median_high``.

HJarausch committed:
Add files via upload

HJarausch committed:
Add files via upload

ghost commented:
I can't comment on if the procs belong to the stats module or not, but please make your code adhere to https://nim-lang.org/docs/nep1.html - try running nimpretty on it first of all.

Yardanico commented on lib/pure/stats.nim:
```diff
@@ -261,6 +292,123 @@ proc kurtosisS*[T](x: openArray[T]): float =
+  assert(left<=K and K<=right)
+  var pInd = K  # Pick `pIndex` as a pivot from the list
+  let pivot = A[pInd]
+  if likely(pInd != right) : swap(A[pInd],A[right])  # Move pivot to end
```
`likely` should IMO be removed, it has been discussed a few times already

Yardanico commented on lib/pure/stats.nim:
```diff
@@ -261,6 +292,123 @@ proc kurtosisS*[T](x: openArray[T]): float =
+    
+    >>> **A** will be modified
+    
+  Timings for random arrays on an amd64 / 4GHz machine with  nim c -d:release -d:danger -z 
```
Not sure about this - do we want to provide some benchmark results right in the documentation?

HJarausch replied:
This mini benchmark should only show that the procedure is very fast. It can be omitted though.
IMHO, these procs to belong to the stats module. The **median** is a robust estimator of the expectation of a random disctribution.

In Python we have  _from statistics import median_,

timotheecour replied:
add benchmark tests in `tests/benchmarks/tstats.nim` (and see the readme tests/benchmarks/readme.md) so that people can re-run it.

and compare it to a naive approach based on:
```nim
proc selectNaive[T](a: seq[T], index: int): T =
  a.sort()[index]
```
(we should also have in future work a `sortN` which returns sorted results for bottom N items; easy to do at least with quicksort; and then in future work your API can be compared to:
```nim
proc selectNaive2[T](a: seq[T], index: int): T =
  a.sortN(index)[index]
```

ringabout commented:
see also our contributing guide
https://nim-lang.github.io/Nim/contributing.html

timotheecour commented on lib/pure/stats.nim:
```diff
@@ -53,6 +53,37 @@ runnableExamples:
 
+  var 
+    ASv = [7,4,6,3,9,1]        #  sorted   1 3 4 6 7 9
+    A   : array[len(ASv),int]
```
follow nep1 styleguide + read contributing.html

konsumlamm commented on lib/pure/stats.nim:
```diff
@@ -53,6 +53,37 @@ runnableExamples:
+  func myCmp(x,y:int):int =
+    if x==y : 0 elif x<y : -1 else: 1
+
+  A= ASv;  doAssert quickSelect(A, 2, 5, 2, myCmp) == 1  # smallest element
+  A= ASv;  doAssert quickSelect(A, 2, 5, 5, myCmp) == 9  # largest element
+  #
+  A= ASv;  doAssert median(A,0,5,myCmp)      == 5.0
+  A= ASv;  doAssert median_low(A,0,5,myCmp)  == 4
+  A= ASv;  doAssert median_high(A,0,5,myCmp) == 6
+  #
+  A= ASv;  doAssert median(A,2,5)            == 4.5
+  A= ASv;  doAssert median_low(A,2,5)        == 3
+  A= ASv;  doAssert median_high(A,2,5)       == 6
+  #
+  A= ASv;  doAssert median(A,0,4,myCmp)      == 6.0
+  A= ASv;  doAssert median_low(A,0,4,myCmp)  == 6
+  A= ASv;  doAssert median_high(A,0,4,myCmp) == 6
```
This is too much for the module level `runnableExamples` imo, it should go to the documentation of the respective functions or tests.

HJarausch replied:
Done

konsumlamm commented on lib/pure/stats.nim:
```diff
@@ -261,6 +292,123 @@ proc kurtosisS*[T](x: openArray[T]): float =
   rs.push(x)
   result = rs.kurtosisS()
 
+import random
```
```suggestion
import std/random
```

Araq replied:
Inside `std` (and lib/pure is std too) imports shouldn't use `std/`

HJarausch replied:
Done

konsumlamm commented on lib/pure/stats.nim:
```diff
@@ -261,6 +292,123 @@ proc kurtosisS*[T](x: openArray[T]): float =
+    return quickSelect(A, pInd + 1, right, K, myCmp)
+
+
+proc median_low*[T](A: var openArray[T], left, right : Natural, 
```
```suggestion
proc medianLow*[T](A: var openArray[T], left, right : Natural, 
```

HJarausch replied:
Done, though I don't like camelCase here.

Yardanico replied:
@HJarausch camelCase is the preferred style for naming in Nim - stdlib should use camelCase whether possible, users of the stdlib can use snake_case in their code if they want to.

Yardanico replied:
Also @HJarausch I don't see any of the changes as additional commits in this PR, did you push it yet?

HJarausch replied:
Done!

konsumlamm commented on lib/pure/stats.nim:
```diff
@@ -261,6 +292,123 @@ proc kurtosisS*[T](x: openArray[T]): float =
+  quickSelect(A, left,right,(right+left) div 2, myCmp)
+
+
+proc median_high*[T](A: var openArray[T], left, right : Natural, 
```
```suggestion
proc medianHigh*[T](A: var openArray[T], left, right : Natural, 
```

timotheecour commented on lib/pure/stats.nim:
```diff
@@ -53,6 +53,37 @@ runnableExamples:
+    A   : array[len(ASv),int]
+
+  func myCmp(x,y:int):int =
+    if x==y : 0 elif x<y : -1 else: 1
```
the only other option is to add those APIs it in std/algorithm instead of std/stats; let's think a bit more which is best and let others chime in.

HJarausch replied:
The proc ``quickSelect`` is a relative of the ``quicksort`` algorithm. So, ``quickSelect`` could be a part of std/algorithm.
But, ``median`` is a well-known mathematical function from statistics.

timotheecour commented on lib/pure/stats.nim:
```diff
@@ -53,6 +53,37 @@ runnableExamples:
+  #
+  A= ASv;  doAssert median(A,0,5,myCmp)      == 5.0
+  A= ASv;  doAssert median_low(A,0,5,myCmp)  == 4
+  A= ASv;  doAssert median_high(A,0,5,myCmp) == 6
```
expose only a single API `median` with a param medianMode, an enum

HJarausch replied:
That's difficult. 
```nim
proc median*[T: SomeNumber]
```
return a **float** value for all ``T`` while   ``median_low[T]/median_high[T]`` return a value of type ``T``.
``median_low/median_high`` could be joined into a single proc - but  would should it be called?

HJarausch committed:
Adapt tstats to new API of median

HJarausch committed:
incorporated suggested changes

Code made conforming to NEP1.
Removed ``likely``.
Changed API to a single (overloaded) ``median`` procedure.

HJarausch committed:
Tiny benchmark for ``quickSelect``

timotheecour commented on lib/pure/stats.nim:
```diff
@@ -53,36 +53,17 @@ runnableExamples:
-  doAssert median(F,0,5,FCmp) == 5.0
+  block:
+    var
+      ASv = [4, 6, 3, 9, 1] #  sorted   1 3 4 6 9
```
```suggestion
      ASv = [4, 6, 3, 9, 1]
      assert ASv.sorted == [1 3 4 6 9]
```
prefer checked assertions to code comments
(add ``from std/algorithm import sorted` on top of runnableExamples if needed)

HJarausch replied:
I don't think ``assert`` for a builtin proc (``sort``) makes sense here. The comment was just an aid for a human reader to check the assertions below.

timotheecour commented on lib/pure/stats.nim:
```diff
@@ -53,36 +53,17 @@ runnableExamples:
+    func myCmp(x, y: int): int =
+      if x == y: 0 elif x < y: -1 else: 1
+
+    A = ASv; doAssert median(A, 1, 4) == 4.5
```
```suggestion
    a = aSv
    doAssert median(A, 1, 4) == 4.5
```

timotheecour commented on lib/pure/stats.nim:
```diff
@@ -53,36 +53,17 @@ runnableExamples:
+  block:
+    var
+      ASv = [4, 6, 3, 9, 1] #  sorted   1 3 4 6 9
+      A: array[len(ASv), int]
```
see nep1: variables are lowercase

timotheecour commented on lib/pure/stats.nim:
```diff
@@ -309,104 +290,102 @@ proc partition[T](A: var openArray[T], left, right, K : Natural,
-                    myCmp : proc(x,y:T):int {.nimcall} = cmp[T]): float =
-  ##[This is a special version which returns a float value. 
-    If the length (right+1-left) is *even*, it returns (median_low + median_high)*0.5
+proc median*[T: SomeNumber](A: var openArray[T], left, right: Natural,
```
what's the point of `left` and `right`?

this works:
```nim
proc foo[T](a: var openArray[T]) =
  echo a
var a = @[10,11,12,13]
foo(a.toOpenArray(1, 2))
```
it's more composable.

HJarausch replied:
Thanks! I didn't know one can pass the result of ``toOpenArray`` to a **var** openArray parameter

timotheecour commented on lib/pure/stats.nim:
```diff
@@ -309,104 +290,102 @@ proc partition[T](A: var openArray[T], left, right, K : Natural,
+    If the length (right+1-left) is *even*,
+    it returns (medianLow + medianHigh)*0.5
 
   >>> **A** will be modified
```
use admonition which renders better:
```suggestion
  ## .. warning:: `A` will be modified.
```
(also, can you make it more precise? eg `A` will be modified if unsorted?)

timotheecour commented on tests/stdlib/tstats.nim:
```diff
@@ -30,6 +30,38 @@ rs1.push(@[1.0, 2.2, 1.4, 4.9])
+  A= ASv;  doAssert median(A,0,4,myCmp) == 6.0
+  A= ASv;  doAssert median(A,0,4,mdlow,myCmp) == 6
+  A= ASv;  doAssert median(A,0,4,mdhigh,myCmp) == 6
+
```
lots of nep1 and contributing.html violations; please read nep1 and recently written tests for best practices.

note that the PR itself is appreciated.

HJarausch replied:
Thanks for your careful proof-reading of this PR and your valuable suggestions.
I hope I have implemented these correctly.

HJarausch committed:
incorporate the changes suggested by Timothee

HJarausch committed:
adapt to the current stats.nim

HJarausch committed:
adapt to the current version of stats.nim

HJarausch committed:
adapt to the recent version of stats.nim

HJarausch committed:
upload correct version of tstats.nim

codic12 commented:
What's blocking this issue? Looks like some CI is failing :thinking: 
https://github.com/nim-lang/Nim/pull/17756/checks?check_run_id=2392184550 not sure if these are related though.

timotheecour commented:
> What's blocking this issue? Looks like some CI is failing 🤔
> #17756 (checks) not sure if these are related though.

tchannels timeout is unrelated, but that is fixed in devel please rebase (also to avoid merge conflicts)

ErikSchierboom commented:
I'd love to see this included!

mikko-m commented:
This would be very valuable contribution, now I personally resorted to copy-pasting the code on this PR to my own utils file...

mikko-m commented:
One VERY important point for this median implementation - it has side effects ie. it MODIFIES the original array it's given. This should be very clearly stated as one does not assume side effects to the original array from median calculation..
